File: examples/weather-mcp/src/test/java/com/example/weather/WeatherServerTest.java
/*
 * Copyright (c) 2026 Konstantin Pavlov/IT Staff and contributors.
 */

package com.example.weather;

import com.example.weather.service.WeatherService;
import dev.tachyonmcp.core.server.TachyonServer;
import io.modelcontextprotocol.client.McpClient;
import io.modelcontextprotocol.client.McpSyncClient;
import io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport;
import io.modelcontextprotocol.spec.McpError;
import io.modelcontextprotocol.spec.McpSchema;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.Assertions.tuple;
import static org.awaitility.Awaitility.await;

class WeatherServerTest {

    private static final TestWeatherProvider weatherProvider = new TestWeatherProvider();
    private static final TestCityProvider cityProvider = new TestCityProvider();
    private static final WeatherService weatherService = new WeatherService(weatherProvider, cityProvider);
    private static TachyonServer handle;
    private static HttpClientStreamableHttpTransport clientTransport;
    private static McpSyncClient client;
    private static McpSchema.InitializeResult initResult;

    @BeforeAll
    static void beforeAll() {
        handle = WeatherServer.buildServer("localhost", 0, null, weatherService);
        handle.start();
        int port = handle.port();

        clientTransport = HttpClientStreamableHttpTransport
            .builder("http://localhost:" + port)
            .build();
        client = McpClient.sync(clientTransport)
            .elicitation(request -> new McpSchema.ElicitResult(
                McpSchema.ElicitResult.Action.ACCEPT, Map.of("city", "Tallinn")))
            .build();

        initResult = client.initialize();
    }

    @AfterAll
    static void afterAll() {
        if (client != null) {
            client.close();
        }
        if (clientTransport != null) {
            clientTransport.close();
        }
        if (handle != null) {
            handle.close();
        }
    }

    @Test
    void shouldGetServerInfo() {
        assertThat(initResult.serverInfo()).usingRecursiveComparison().ignoringFields("icons").isEqualTo(
            McpSchema.Implementation.builder("weather-server", "1.0")
                .title("Weather Server")
                .websiteUrl("https://github.com/tachyonmcp/tachyon/tree/main/examples/weather-mcp")
                .description("Weather MCP server")
                .build());
        assertThat(initResult.serverInfo().icons()).singleElement().satisfies(icon -> {
            assertThat(icon.src()).startsWith("data:image/png;base64,");
            assertThat(icon.mimeType()).isEqualTo("image/png");
            assertThat(icon.sizes()).containsExactly("256x256");
        });
        assertThat(initResult.protocolVersion()).isEqualTo("2025-11-25");
        assertThat(initResult.instructions()).isEqualTo("Test instructions");
        assertThat(initResult.capabilities()).isEqualTo(McpSchema.ServerCapabilities.builder()
            .tools(null)
            .resources(null, null)
            .prompts(null)
            .logging()
            .completions()
            .build());
    }

    @Test
    void shouldListTools() {
        final var result = client.listTools();
        assertThat(result).isNotNull();
        assertThat(result.tools()).hasSize(1);
        McpSchema.Tool tool = result.tools().getFirst();
        assertThat(tool.name()).isEqualTo("get-weather");
        assertThat(tool.title()).isEqualTo("Current Weather");
        assertThat(tool.description()).isEqualTo("Get current weather for a city");
        assertThat(tool.inputSchema()).isEqualTo(Map.of(
            "$schema", "https://json-schema.org/draft/2020-12/schema",
            "$id", "GetWeatherRequest",
            "description", "Input for looking up the current weather in a city.",
            "type", "object",
            "required", List.of("city", "units"),
            "additionalProperties", false,
            "properties", Map.of(
                "city", Map.of(
                    "description", "City name (e.g., London, Tokyo, New York)",
                    "type", "string"),
                "units", Map.of(
                    "description", "Temperature unit (default: celsius)",
                    "oneOf", List.of(
                        Map.of("type", "null"),
                        Map.of("$ref", "#/$defs/TemperatureUnit")))),
            "$defs", Map.of(
                "TemperatureUnit", Map.of(
                    "type", "string",
                    "description", "Unit used to represent temperature.",
                    "enum", List.of("celsius", "fahrenheit")))));
        assertThat(tool.outputSchema()).isEqualTo(Map.of(
            "$schema", "https://json-schema.org/draft/2020-12/schema",
            "$id", "GetWeatherResponse",
            "description", "Current weather observation for a city.",
            "type", "object",
            "required", List.of("city", "condition", "temperature", "unit", "humidity", "windSpeed"),
            "additionalProperties", false,
            "properties", Map.of(
                "city", Map.of("description", "City name", "type", "string"),
                "condition", Map.of("description", "Weather condition", "type", "string"),
                "temperature", Map.of("description", "Temperature in the response unit", "type", "number"),
                "unit", Map.of("description", "Temperature unit", "$ref", "#/$defs/TemperatureUnit"),
                "humidity", Map.of("description", "Relative humidity percentage", "type", "integer"),
                "windSpeed", Map.of("description", "Wind speed in km/h", "type", "number")),
            "$defs", Map.of(
                "TemperatureUnit", Map.of(
                    "type", "string",
                    "description", "Unit used to represent temperature.",
                    "enum", List.of("celsius", "fahrenheit")))));
        assertThat(tool.meta()).isNull();
    }

    @Test
    void shouldCallWeatherTool() {
        final var result = client.callTool(McpSchema.CallToolRequest.builder("get-weather")
            .arguments(Map.of("city", "London", "units", "celsius"))
            .build());

        assertThat(result).isNotNull();
        assertThat(result.structuredContent()).isEqualTo(Map.of(
            "city", "London",
            "condition", "Clear sky",
            "temperature", 18.5,
            "unit", "celsius",
            "humidity", 52,
            "windSpeed", 12.0));
    }

    @Test
    void shouldEmitProgressWhileFetchingWeather() {
        var progressNotifications = new CopyOnWriteArrayList<McpSchema.ProgressNotification>();
        var progressScheduler = Schedulers.newSingle("weather-progress");
        var progressClient = McpClient.async(HttpClientStreamableHttpTransport
            .builder("http://localhost:" + handle.port())
            .build())
            .progressConsumer(notification -> Mono.<Void>fromRunnable(() -> progressNotifications.add(notification))
                .subscribeOn(progressScheduler))
            .build();
        try {
            progressClient.initialize().block();
            var arguments = new HashMap<String, Object>();
            arguments.put("city", "London");
            arguments.put("units", null);
            final var result = progressClient.callTool(McpSchema.CallToolRequest.builder("get-weather")
                .arguments(arguments)
                .progressToken("weather-progress")
                .build())
                .block();

            assertThat(result.isError()).isNotEqualTo(true);
            await().atMost(Duration.ofSeconds(5)).untilAsserted(() ->
                assertThat(progressNotifications)
                    .extracting(
                        McpSchema.ProgressNotification::progressToken,
                        McpSchema.ProgressNotification::progress,
                        McpSchema.ProgressNotification::total,
                        McpSchema.ProgressNotification::message)
                    .containsExactly(
                        tuple("weather-progress", 0.1, 1.0, "Fetching weather for London"),
                        tuple("weather-progress", 1.0, 1.0, "Weather retrieved for London")));
        } finally {
            progressClient.close();
            progressScheduler.dispose();
        }
    }

    @Test
    void shouldCallWeatherToolAfterElicitingAnotherCity() {
        var arguments = new HashMap<String, Object>();
        arguments.put("city", "Unknown");
        arguments.put("units", null);
        final var result = client.callTool(McpSchema.CallToolRequest.builder("get-weather")
            .arguments(arguments)
            .build());

        assertThat(result.structuredContent()).isEqualTo(Map.of(
            "city", "Tallinn",
            "condition", "Clear sky",
            "temperature", 18.5,
            "unit", "celsius",
            "humidity", 52,
            "windSpeed", 12.0));
    }

    @Test
    void shouldListResources() {
        final var result = client.listResources();

        assertThat(result.resources()).hasSize(2);

        var article = result.resources().stream()
            .filter(resource -> resource.uri().equals("weather://prediction/article"))
            .findFirst().orElseThrow();
        assertThat(article.uri()).isEqualTo("weather://prediction/article");
        assertThat(article.name()).isEqualTo("prediction-article");
        assertThat(article.title()).isEqualTo("Weather Prediction");
        assertThat(article.description()).isEqualTo("Weather prediction article");
        assertThat(article.mimeType()).isEqualTo("text/markdown");
        assertThat(article.size()).isEqualTo(weatherService.predictionArticle().getBytes(UTF_8).length);
        assertThat(article.annotations().audience())
            .containsExactly(McpSchema.Role.USER, McpSchema.Role.ASSISTANT);
        assertThat(article.annotations().priority()).isEqualTo(0.8);
        assertThat(article.annotations().lastModified()).isEqualTo("2026-07-23T00:00:00Z");
        assertThat(article.icons()).singleElement().satisfies(icon -> {
            assertThat(icon.src()).startsWith("data:image/png;base64,");
            assertThat(icon.mimeType()).isEqualTo("image/png");
            assertThat(icon.sizes()).containsExactly("256x256");
            assertThat(icon.theme()).isEqualTo("light");
        });

        var weather = result.resources().stream()
            .filter(resource -> resource.uri().equals("weather://featured/current"))
            .findFirst().orElseThrow();
        assertThat(weather.uri()).isEqualTo("weather://featured/current");
        assertThat(weather.name()).isEqualTo("featured-current-weather");
        assertThat(weather.title()).isEqualTo("Featured Current Weather");
        assertThat(weather.description()).isEqualTo("Current weather in Tallinn");
        assertThat(weather.mimeType()).isEqualTo("application/json");
        assertThat(weather.annotations()).isEqualTo(article.annotations());
        assertThat(weather.icons()).isEqualTo(article.icons());
    }

    @Test
    void shouldReadTextResource() {
        final var listResult = client.listResources();
        var article = listResult.resources().stream()
            .filter(r -> r.uri().equals("weather://prediction/article"))
            .findFirst().orElseThrow();

        final var result = client.readResource(article);

        var contents = result.contents().getFirst();
        assertThat(contents).isInstanceOf(McpSchema.TextResourceContents.class);
        var textContents = ((McpSchema.TextResourceContents) contents);
        assertThat(textContents.uri()).isEqualTo("weather://prediction/article");
        assertThat(textContents.mimeType()).isEqualTo("text/markdown");
        assertThat(textContents.text().trim())
            .startsWith("# Weather Prediction")
            .endsWith("reports, ocean buoys, and over 30 polar-orbiting and geostationary satellites.");
    }

    @Test
    void shouldReadCurrentWeatherResource() {
        final var listResult = client.listResources();
        var weather = listResult.resources().stream()
            .filter(r -> r.uri().equals("weather://featured/current"))
            .findFirst().orElseThrow();

        final var result = client.readResource(weather);

        var contents = result.contents().getFirst();
        assertThat(contents).isInstanceOf(McpSchema.TextResourceContents.class);
        var textContents = ((McpSchema.TextResourceContents) contents);
        assertThat(textContents.uri()).isEqualTo("weather://featured/current");
        assertThat(textContents.mimeType()).isEqualTo("application/json");
        assertThat(textContents.text()).contains("Clear sky");
    }

    @Test
    void shouldListResourceTemplates() {
        final var result = client.listResourceTemplates();

        assertThat(result.resourceTemplates()).hasSize(1);
        var template = result.resourceTemplates().getFirst();
        assertThat(template.uriTemplate()).isEqualTo("weather://current/{city}");
        assertThat(template.name()).isEqualTo("current-weather");
        assertThat(template.mimeType()).isEqualTo("application/json");
    }

    @Test
    void shouldReadCurrentWeatherFromTemplate() {
        final var result = client.readResource(
            McpSchema.ReadResourceRequest.builder("weather://current/London").build());

        var contents = result.contents().getFirst();
        assertThat(contents).isInstanceOf(McpSchema.TextResourceContents.class);
        var textContents = ((McpSchema.TextResourceContents) contents);
        assertThat(textContents.uri()).isEqualTo("weather://current/London");
        assertThat(textContents.mimeType()).isEqualTo("application/json");
        assertThat(textContents.text()).contains("Clear sky");
    }

    @Test
    void shouldReturnInvalidParamsWhenTemplateCityIsUnknown() {
        assertThatThrownBy(() -> client.readResource(
                McpSchema.ReadResourceRequest.builder("weather://current/Unknown").build()))
            .isInstanceOf(McpError.class)
            .extracting(e -> ((McpError) e).getJsonRpcError().code())
            .isEqualTo(-32602);
    }

    @Test
    void shouldCompleteCityNameForCurrentWeatherTemplate() {
        final var result = client.completeCompletion(McpSchema.CompleteRequest.builder(
                new McpSchema.ResourceReference("weather://current/{city}"),
                new McpSchema.CompleteRequest.CompleteArgument("city", "Lo"))
            .build());

        assertThat(result.completion().values()).containsExactlyInAnyOrder("London", "Los Angeles");
        assertThat(result.completion().hasMore()).isNotEqualTo(true);
    }

    @Test
    void shouldReturnEmptyCompletionForBlankQuery() {
        final var result = client.completeCompletion(McpSchema.CompleteRequest.builder(
                new McpSchema.ResourceReference("weather://current/{city}"),
                new McpSchema.CompleteRequest.CompleteArgument("city", ""))
            .build());

        assertThat(result.completion().values()).isEmpty();
    }

    @Test
    void shouldCompleteStyleNameForRewriteForecastPrompt() {
        final var result = client.completeCompletion(McpSchema.CompleteRequest.builder(
                new McpSchema.PromptReference("rewrite-forecast"),
                new McpSchema.CompleteRequest.CompleteArgument("style", "pi"))
            .build());

        assertThat(result.completion().values()).containsExactly("pirate");
    }

    @Test
    void shouldReturnEmptyCompletionForNonStyleArgumentOfRewriteForecastPrompt() {
        final var result = client.completeCompletion(McpSchema.CompleteRequest.builder(
                new McpSchema.PromptReference("rewrite-forecast"),
                new McpSchema.CompleteRequest.CompleteArgument("forecast", "Rain"))
            .build());

        assertThat(result.completion().values()).isEmpty();
    }

    @Test
    void shouldListPrompts() {
        final var result = client.listPrompts();

        assertThat(result.prompts()).hasSize(1);
        var prompt = result.prompts().getFirst();
        assertThat(prompt.name()).isEqualTo("rewrite-forecast");
        assertThat(prompt.description()).isEqualTo("Rewrites a weather forecast in a chosen style");
        assertThat(prompt.arguments()).hasSize(2);
    }

    @Test
    void shouldGetPrompt() {
        final var result = client.getPrompt(
            McpSchema.GetPromptRequest.builder("rewrite-forecast")
                .arguments(Map.of("forecast", "Rain in London", "style", "pirate"))
                .build());

        assertThat(result).isNotNull();
        assertThat(result.messages()).hasSize(1);
        var message = result.messages().getFirst();
        assertThat(message.role()).isEqualTo(McpSchema.Role.USER);
        assertThat(message.content()).isInstanceOf(McpSchema.TextContent.class);
        var textContent = ((McpSchema.TextContent) message.content());
        assertThat(textContent.text())
            .isEqualTo("Rewrite the following weather forecast in pirate style. Preserve factual details:\n\n```Rain in London\n```");
    }
}

File: examples/weather-mcp-kotlin/src/test/kotlin/com/example/weather/WeatherServerTest.kt
/*
 * Copyright (c) 2026 Konstantin Pavlov/IT Staff and contributors.
 */

package com.example.weather

import com.example.weather.service.WeatherService
import dev.tachyonmcp.core.server.TachyonServer
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldStartWith
import io.kotest.matchers.types.shouldBeInstanceOf
import io.modelcontextprotocol.client.McpClient
import io.modelcontextprotocol.client.McpSyncClient
import io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport
import io.modelcontextprotocol.spec.McpError
import io.modelcontextprotocol.spec.McpSchema.CallToolRequest
import io.modelcontextprotocol.spec.McpSchema.CompleteRequest
import io.modelcontextprotocol.spec.McpSchema.ElicitResult
import io.modelcontextprotocol.spec.McpSchema.GetPromptRequest
import io.modelcontextprotocol.spec.McpSchema.InitializeResult
import io.modelcontextprotocol.spec.McpSchema.ProgressNotification
import io.modelcontextprotocol.spec.McpSchema.PromptReference
import io.modelcontextprotocol.spec.McpSchema.ReadResourceRequest
import io.modelcontextprotocol.spec.McpSchema.ResourceReference
import io.modelcontextprotocol.spec.McpSchema.Role
import io.modelcontextprotocol.spec.McpSchema.TextContent
import io.modelcontextprotocol.spec.McpSchema.TextResourceContents
import org.awaitility.kotlin.atMost
import org.awaitility.kotlin.await
import org.awaitility.kotlin.untilAsserted
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.time.Duration
import java.util.concurrent.CopyOnWriteArrayList

class WeatherServerTest {
    companion object {
        private val weatherProvider = TestWeatherProvider()
        private val cityProvider = TestCityProvider()
        private val weatherService = WeatherService(weatherProvider, cityProvider)
        private lateinit var server: TachyonServer
        private lateinit var clientTransport: HttpClientStreamableHttpTransport
        private lateinit var client: McpSyncClient
        private lateinit var initResult: InitializeResult
        private val progressNotifications = CopyOnWriteArrayList<ProgressNotification>()

        @JvmStatic
        @BeforeAll
        fun beforeAll() {
            server = assembleServer(0, weatherService)
            server.start()
            val port = server.port()

            clientTransport =
                HttpClientStreamableHttpTransport.builder("http://localhost:$port").build()
            client =
                McpClient
                    .sync(clientTransport)
                    .elicitation { _ ->
                        ElicitResult(
                            ElicitResult.Action.ACCEPT,
                            mapOf("city" to "Tallinn"),
                        )
                    }.progressConsumer { progressNotifications.add(it) }
                    .build()

            initResult = client.initialize()
        }

        @JvmStatic
        @AfterAll
        fun afterAll() {
            client.close()
            clientTransport.close()
            server.close()
        }
    }

    @Test
    fun verifyInitResult() {
        with(initResult.serverInfo()) {
            name() shouldBe "weather-server-kotlin"
            title() shouldBe "Weather Server (Kotlin)"
            description() shouldBe "Weather MCP server built with Tachyon Kotlin DSL"
            websiteUrl() shouldBe
                "https://github.com/tachyonmcp/tachyon/tree/main/examples/weather-mcp-kotlin"
            icons() shouldHaveSize 1
            icons().first().src() shouldStartWith "data:image/png;base64,"
        }
        initResult.protocolVersion() shouldBe "2025-11-25"
        initResult.instructions() shouldBe "Test instructions"
    }

    @Test
    fun `should list tools`() {
        val result = client.listTools()

        result.tools() shouldHaveSize 1
        result.tools().first() shouldNotBeNull {
            name() shouldBe "get-weather"
            title() shouldBe "Current Weather"
            description() shouldBe "Get current weather for a city"
            outputSchema() shouldNotBe null
        }
    }

    @Test
    fun `should call weather tool`() {
        val result =
            client.callTool(
                CallToolRequest
                    .builder("get-weather")
                    .arguments(mapOf("city" to "London", "units" to "Celsius"))
                    .build(),
            )

        result.isError shouldNotBe true
        result.structuredContent() shouldBe
            mapOf(
                "city" to "London",
                "condition" to "Clear sky",
                "temperature" to 18.5,
                "temperatureUnit" to "Celsius",
                "humidity" to 52,
                "windSpeed" to 12.0,
            )
        result
            .content()
            .single()
            .shouldBeInstanceOf<TextContent>()
            .text() shouldBe
            """{"city":"London","condition":"Clear sky","temperature":18.5,"temperatureUnit":"Celsius","humidity":52,"windSpeed":12.0}"""
    }

    @Test
    fun `should emit progress while fetching weather`() {
        val result =
            client.callTool(
                CallToolRequest
                    .builder("get-weather")
                    .arguments(mapOf("city" to "London"))
                    .progressToken("weather-progress")
                    .build(),
            )

        result.isError shouldNotBe true
        await atMost Duration.ofSeconds(5) untilAsserted {
            progressNotifications shouldHaveSize 2
        }
        progressNotifications.map { it.progressToken() to it.message() } shouldContainExactly
            listOf(
                "weather-progress" to "Fetching weather for London",
                "weather-progress" to "Weather retrieved for London",
            )
    }

    @Test
    fun `should call weather tool after eliciting another city`() {
        val result =
            client.callTool(
                CallToolRequest
                    .builder("get-weather")
                    .arguments(mapOf("city" to "Unknown"))
                    .build(),
            )

        result.isError shouldNotBe true
        result.structuredContent() shouldBe
            mapOf(
                "city" to "Tallinn",
                "condition" to "Clear sky",
                "temperature" to 18.5,
                "temperatureUnit" to "Celsius",
                "humidity" to 52,
                "windSpeed" to 12.0,
            )
        result
            .content()
            .single()
            .shouldBeInstanceOf<TextContent>()
            .text() shouldBe
            """{"city":"Tallinn","condition":"Clear sky","temperature":18.5,"temperatureUnit":"Celsius","humidity":52,"windSpeed":12.0}"""
    }

    @Test
    fun `should list resources`() {
        val result = client.listResources()

        result.resources() shouldHaveSize 2
        val article = result.resources().first { it.uri() == "weather://prediction/article" }
        with(article) {
            name() shouldBe "prediction-article"
            title() shouldBe "Weather Prediction"
            description() shouldBe "Weather prediction article"
            mimeType() shouldBe "text/markdown"
            size() shouldBe
                weatherService.predictionArticle
                    .toByteArray()
                    .size
                    .toLong()
            with(annotations()) {
                audience() shouldContainExactly
                    listOf(Role.USER, Role.ASSISTANT)
                priority() shouldBe 0.8
                lastModified() shouldBe "2026-07-23T00:00:00Z"
            }
            with(icons().single()) {
                src() shouldStartWith "data:image/png;base64,"
                mimeType() shouldBe "image/png"
                sizes() shouldContainExactly listOf("256x256")
                theme() shouldBe "light"
            }
        }

        val weather =
            result.resources().first {
                it.uri() == "weather://featured/current"
            }
        with(weather) {
            name() shouldBe "featured-current-weather"
            title() shouldBe "Featured Current Weather"
            description() shouldBe "Current weather in Tallinn"
            mimeType() shouldBe "application/json"
            annotations() shouldBe article.annotations()
            icons() shouldBe article.icons()
        }
    }

    @Test
    fun `should read text resource`() {
        val article =
            client.listResources().resources().first { it.uri() == "weather://prediction/article" }

        val result = client.readResource(article)

        val contents = result.contents().first()
        with(contents) {
            shouldBeInstanceOf<TextResourceContents>()
            uri() shouldBe "weather://prediction/article"
            mimeType() shouldBe "text/markdown"
            text().trim() shouldStartWith "# Weather Prediction"
        }
    }

    @Test
    fun `should read current weather resource`() {
        val weather =
            client.listResources().resources().first { it.uri() == "weather://featured/current" }

        val result = client.readResource(weather)

        val contents = result.contents().first()
        with(contents) {
            shouldBeInstanceOf<TextResourceContents>()
            uri() shouldBe "weather://featured/current"
            mimeType() shouldBe "application/json"
            text() shouldContain "Clear sky"
        }
    }

    @Test
    fun `should list resource templates`() {
        val result = client.listResourceTemplates()

        result.resourceTemplates() shouldHaveSize 1
        result.resourceTemplates().first() shouldNotBeNull {
            uriTemplate() shouldBe "weather://current/{city}"
            name() shouldBe "current-weather"
            mimeType() shouldBe "application/json"
        }
    }

    @Test
    fun `should read current weather from template`() {
        val result =
            client.readResource(
                ReadResourceRequest.builder("weather://current/London").build(),
            )

        val contents = result.contents().first()
        contents.shouldBeInstanceOf<TextResourceContents>()
        contents.uri() shouldBe "weather://current/London"
        contents.mimeType() shouldBe "application/json"
        contents.text() shouldContain "Clear sky"
    }

    @Test
    fun `should return invalid params when template city is unknown`() {
        val error =
            shouldThrow<McpError> {
                client.readResource(
                    ReadResourceRequest.builder("weather://current/Unknown").build(),
                )
            }
        error.jsonRpcError.code() shouldBe -32602
    }

    @Test
    fun `should complete city name for current weather template`() {
        val result =
            client.completeCompletion(
                CompleteRequest
                    .builder(
                        ResourceReference("weather://current/{city}"),
                        CompleteRequest.CompleteArgument("city", "Lo"),
                    ).build(),
            )

        result.completion().values() shouldContainExactlyInAnyOrder listOf("London", "Los Angeles")
        result.completion().hasMore() shouldNotBe true
    }

    @Test
    fun `should return empty completion for blank query`() {
        val result =
            client.completeCompletion(
                CompleteRequest
                    .builder(
                        ResourceReference("weather://current/{city}"),
                        CompleteRequest.CompleteArgument("city", ""),
                    ).build(),
            )

        result.completion().values() shouldHaveSize 0
    }

    @Test
    fun `should complete style name for rewrite-forecast prompt`() {
        val result =
            client.completeCompletion(
                CompleteRequest
                    .builder(
                        PromptReference("rewrite-forecast"),
                        CompleteRequest.CompleteArgument("style", "pi"),
                    ).build(),
            )

        result.completion().values() shouldContainExactly listOf("pirate")
    }

    @Test
    fun `should return empty completion for non-style argument of rewrite-forecast prompt`() {
        val result =
            client.completeCompletion(
                CompleteRequest
                    .builder(
                        PromptReference("rewrite-forecast"),
                        CompleteRequest.CompleteArgument("forecast", "Rain"),
                    ).build(),
            )

        result.completion().values() shouldHaveSize 0
    }

    @Test
    fun `should list prompts`() {
        val result = client.listPrompts()

        result.prompts() shouldHaveSize 1
        val prompt = result.prompts().first()
        prompt.name() shouldBe "rewrite-forecast"
        prompt.description() shouldBe "Rewrites a weather forecast in a chosen style"
        prompt.arguments() shouldHaveSize 2
    }

    @Test
    fun `should get prompt`() {
        val result =
            client.getPrompt(
                GetPromptRequest
                    .builder("rewrite-forecast")
                    .arguments(mapOf("forecast" to "Rain in London", "style" to "pirate"))
                    .build(),
            )

        result.messages() shouldHaveSize 1
        val message = result.messages().first()
        message.role() shouldBe Role.USER
        message.content().shouldBeInstanceOf<TextContent>()
        (message.content() as TextContent).text() shouldBe
            "Rewrite the following weather forecast in pirate style. Preserve factual details:\n\n```Rain in London\n```"
    }
}
